src/debug/telemetry/performance_monitor.c

```diff
@@ -397,3 +404,225 @@ int performance_monitor_init(void)
+#define IO_PERFORMANCE_MAX_ENTRIES (IO_PERFORMANCE_ALLOC_BYTES / sizeof(struct io_perf_data_item))
+
+static struct io_perf_monitor_ctx perf_monitor_ctx;
+static struct io_perf_data_item io_perf_data_items[IO_PERFORMANCE_MAX_ENTRIES];
+
+int io_perf_monitor_init(void)
+{
+	int ret;
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+
+	k_spinlock_init(&self->lock);
+	k_spinlock_key_t key = k_spin_lock(&self->lock);
+
+	self->io_perf_data = (struct io_perf_data_item *)io_perf_data_items;
+	self->state = IPC4_PERF_MEASUREMENTS_DISABLED;
+
+	ret = perf_bitmap_init(&self->io_performance_data_bitmap, &io_performance_data_bit_array,
+			       IO_PERFORMANCE_MAX_ENTRIES);
+
+	k_spin_unlock(&self->lock, key);
+	return ret;
+}
+
+/* init performance monitor using Zephyr */
+SYS_INIT(io_perf_monitor_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
+
+static struct io_perf_data_item *io_perf_monitor_get_next_slot(struct io_perf_monitor_ctx *self)
+{
+	int idx;
+	int ret;
+	k_spinlock_key_t key = k_spin_lock(&self->lock);
+
+	ret = perf_bitmap_alloc(&self->io_performance_data_bitmap, &idx);
+	if (ret < 0)
+		return NULL;
+	/* ref. FW did not set the bits, but here we do it to not have to use
+	 * isFree() check that the bitarray does not provide yet. Instead we will use isClear
+	 * ,and always set bit on bitmap alloc.
+	 */
+
+	ret = perf_bitmap_setbit(&self->io_performance_data_bitmap, idx);
+	if (ret < 0)
+		return NULL;
+
+	k_spin_unlock(&self->lock, key);
+	return &self->io_perf_data[idx];
+}
+
+int io_perf_monitor_release_slot(struct io_perf_data_item *item)
+{
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+	int idx;
+	int ret;
+	k_spinlock_key_t key;
+
+	item->is_removed = true;
+
+	key = k_spin_lock(&self->lock);
+	idx = (item - self->io_perf_data) / sizeof(*item);
+	k_spin_unlock(&self->lock, key);
+
+	ret = perf_bitmap_clearbit(&self->io_performance_data_bitmap, idx);
+	if (ret < 0)
+		return ret;
+	ret = perf_bitmap_free(&performance_data_bitmap, idx);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+int
+io_perf_monitor_get_performance_data(struct io_global_perf_data *io_global_perf_data)
+{
+	if (!io_global_perf_data)
+		return 0;
+
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+	k_spinlock_key_t key = k_spin_lock(&self->lock);
+
+	size_t slot_idx = 0;
+	size_t slots_count = self->io_performance_data_bitmap.occupied;
+	size_t entries_cont = self->io_performance_data_bitmap.size;
+
+	for (size_t idx = 0; idx < entries_cont && slot_idx < slots_count; ++idx) {
+		if (perf_bitmap_is_bit_clear(&self->io_performance_data_bitmap, idx))
+			continue;
+		io_global_perf_data->perf_items[slot_idx] = self->io_perf_data[idx];
+		++slot_idx;
+	}
+	io_global_perf_data->perf_item_count = slots_count;
+
+	k_spin_unlock(&self->lock, key);
+	return 0;
+}
+
+static int
+io_perf_monitor_disabled(struct io_perf_monitor_ctx *self)
+{
+	return 0;
+}
+
+static int
+io_perf_monitor_stopped(struct io_perf_monitor_ctx *self)
```

Comment:
this never returns errors, so can be `void`. Besides the name sounds like it's checking whether the monitor has stopped, while this isn't what this function is doing